main/src/library/Set.flix

```diff
@@ -328,23 +316,55 @@ namespace Set {
     /// make any guarantees about which mapping will be in the resulting map.
     ///
     pub def toMap(xs: Set[(a, b)]): Map[a, b] =
-        let Set(s) = xs;
-            List.toMap(s)
+        foldRight((x, acc) -> let (k, v) = x; Map.insert(k, v, acc), Map#{}, xs)
 
     ///
     /// Applies `f` to every element of `xs`.
     ///
     /// The function `f` must be impure.
     ///
-    pub def foreach(f: a ~> Unit, xs: Set[a]): Unit & Impure = match xs {
-        case Set(l) => List.foreach(f, l)
+    pub def foreach(f: a ~> Unit, xs: Set[a]): Unit & Impure =
+        let Set(es) = xs;
+        RBTree.foreach((x, _) -> f(x), es)
+
+    ///
+    /// Helper function for `cmp`.
+    ///
+    def __cmpHelper(xs: List[a], ys: List[a]): Int32 = match (xs, ys) {
+        case (z :: zs, w :: ws) =>
+            let cmp = z <=> w;
+            if (cmp == 0) __cmpHelper(zs, ws) else cmp
+        case (_ :: _, Nil) => 1
+        case (Nil, _ :: _) => -1
+        case (Nil, Nil) => 0
     }
 
+    ///
+    /// Returns the result of a three-way comparison between `xs` and `ys`.
+    ///
+    /// Returns n < 0 if `xs` is lexicographically smaller than `ys`.
+    /// Returns n = 0 if `xs` is equal to `ys`.
+    /// Returns n > 0 if `xs` is lexicographically greater than `ys`.
+    ///
+    pub def __cmp(xs: Set[a], ys: Set[a]): Int32 =
+        __cmpHelper(toList(xs), toList(ys))
 
+    ///
+    /// Helper function for `range`.
+    ///
+    def rangeHelper(b: Int32, n: Int32, acc: Set[Int32]): Set[Int32] =
+        if (n < b) acc else rangeHelper(b, n - 1, insert(n, acc))
 
+    ///
+    /// Returns a set of all integers between `b` (inclusive) and `e` (exclusive).
+    ///
+    /// Returns `empty()` if `b >= e`.
+    ///
+    pub def range(b: Int32, e: Int32): Set[Int32] =
```

Comment:
Move to range function a bit upwards, see if you can find some place it "fits in".